### PR TITLE
Introduce getting holds and listing bookmarks through ZCP

### PR DESCRIPTION
**Pull request by sdimitro**
Consumers of ZFS Channel Programs can now list bookmarks,
and get holds from datasets. A minor-refactoring was also
applied to distinguish between user and system properties
in ZCP.

### How Has This Been Tested?

New tests added to ZTS.
Has been in DelphixOS(illumos) for more than a year.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [X] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by sdimitro** [comment]
These changes are based on top of redacted send receive and should be open for a clean review once that lands first.

File: module/zfs/zcp_iter.c [hunk 1]
```diff
@@ -361,11 +360,11 @@ zcp_props_iter(lua_State *state)
...
 }
 
-static int zcp_props_list(lua_State *);
-static zcp_list_info_t zcp_props_list_info = {
-	.name = "properties",
-	.func = zcp_props_list,
-	.gc = zcp_props_list_gc,
+static int zcp_user_props_list(lua_State *);
+static zcp_list_info_t zcp_user_props_list_info = {
+	.name = "user_properties",
```
Comment (ahrens) [inline module/zfs/zcp_iter.c]: are we concerned about changing the name?  Maybe not if `"properties"` isn't part of a release yet?

Comment (sdimitro) [reply module/zfs/zcp_iter.c]: So this code was introduced with the initial channel program commit -> d99a015343425a1c856c900aa8223016400ac2dc

so I assume it has been part of a release. @tonyhutter is this right?
If it is what do you think would be the impact of this change would be?
(e.g. maybe distro packages of zfs are so behind that hadn't had the time to use this feature in LTS releases, etc...)

Comment (tonyhutter) [reply module/zfs/zcp_iter.c]: Channel programs arn't in the 0.7.x point releases, so you're free to change the name.

**Comment by ahrens** [comment]
I think this is ready to go, but it's labeled "work in progress".  Is there something else we're looking for here?  I checked one of the failed tests and it has: 
```
Tests with results other than PASS that are unexpected:
    FAIL cli_root/zfs_receive/zfs_receive_raw_incremental (expected PASS)
    FAIL cli_root/zfs_rename/zfs_rename_004_neg (expected PASS)
    FAIL cli_root/zfs_send/zfs_send-b (expected PASS)
    FAIL cli_root/zpool_get/zpool_get_002_pos (expected PASS)
    FAIL libzfs/libzfs_input (expected PASS)
    FAIL rsend/send-cD (expected PASS)
```
Which I'm not sure if they are related to this change.

**Comment by sdimitro** [comment]
@ahrens This work is based on top of Redacted Send/Receive so that should go in first before this one is applied.

**Comment by sdimitro** [comment]
Now that Redacted Send/Receive is in, we should get this in. I'm rebasing this work to rerun the tests

**Comment by sdimitro** [comment]
@ahrens @pcd this looks like it can get reviewed now that redacted send/receive is in and the tests pass.

File: module/zfs/zcp_iter.c (hunk 1 above)
Comment (ahrens) [reply module/zfs/zcp_iter.c]: Channel programs are now in 0.8 release, so are we concerned about changing this name?

Comment (sdimitro) [reply module/zfs/zcp_iter.c]: @tonyhutter @behlendorf  I'm ok reverting the renaming, just wanted to get your thoughts before doing so.

Comment (sdimitro) [reply module/zfs/zcp_iter.c]: Ended up leaving `user_properties` as is and making `properties` an alias to it. Added a comment about this decision above the declaration of `zcp_props_list_info`, and a mention of the alias in the man page. I've also updated their respective tests, to test both APIs. Let me know what you think.

Comment (behlendorf) [reply module/zfs/zcp_iter.c]: Including a `properties` alias for `user_properties` sounds like a good solution to me.

File: tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.list_bookmarks.ksh [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	return 0
+EOF
+log_mustnot_program $TESTPOOL - <<-EOF
+	zfs.list.bookmarks("nonexistent-pool/$TESTFS")
+	return 0
+EOF
+
+# Can't look in a different pool than the one specified on command line
+log_mustnot_program $TESTPOOL - <<-EOF
+	zfs.list.bookmarks("rpool")
```
Comment (pcd1193182) [inline tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.list_bookmarks.ksh]: We're not guaranteed that rpool exists at all; is there a `testpool2` we could use here instead?

File: tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.list_bookmarks.ksh [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	return 0
+EOF
+
+# Can't have bookmarks on snapshots, only on filesystems
+log_mustnot_program $TESTPOOL - <<-EOF
+	zfs.list.bookmarks("$TESTPOOL/$TESTFS@$TESTSNAP")
```
Comment (pcd1193182) [inline tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.list_bookmarks.ksh]: Would this be a possible future enhancement? You could filter by the txg of the bookmark to find ones that match the snapshot.

File: tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.list_holds.ksh [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	return 0
+EOF
+log_mustnot_program $TESTPOOL - <<-EOF
+	zfs.list.holds("nonexistent-pool/$TESTFS")
+	return 0
+EOF
+
+# Can't look in a different pool than the one specified on command line
+log_mustnot_program $TESTPOOL - <<-EOF
+	zfs.list.holds("rpool")
```
Comment (pcd1193182) [inline tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.list_holds.ksh]: See above re: 'testpool2'

File: tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.list_bookmarks.ksh (hunk 2 above)
Comment (sdimitro) [reply tests/zfs-tests/tests/functional/channel_program/synctask_core/tst.list_bookmarks.ksh]: Added `testpool2` at setup and cleanup

**Pull request merged**